### PR TITLE
[tune] Add `set_max_concurrency` to Searcher API

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Adds a `set_max_concurrency` method to the Searcher API. This method allows for the `ConcurrencyLimiter` to override the `max_concurrency` value on searchers with custom internal logic for limiting concurrency (atm. SigOpt and HEBO). This PR also changes the initialisation of SigOpt and HEBO optimisers to happen on `set_search_properties` instead of in the constructor, so that the new `max_concurrency` is respected.

Furthermore, this PR breaks up `test_tune_restore.py` into `test_tune_restore_warm_start.py` and `test_tune_restore.py` to deal with a timeout, and ensures that the automatic application of `ConcurrencyLimiter` in `tune.run` doesn't override a user-defined `ConcurrencyLimiter`.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/suggest/suggestion.py [hunk 1]
```diff
@@ -382,7 +406,25 @@ def __init__(self,
         super(ConcurrencyLimiter, self).__init__(
             metric=self.searcher.metric, mode=self.searcher.mode)
 
+    def set_max_concurrency(self, max_concurrent: int) -> bool:
+        # Determine if this behavior is acceptable, or if it should
+        # raise an exception.
+        self.max_concurrent = max_concurrent
+        return True
```
Comment (Yard1) [inline python/ray/tune/suggest/suggestion.py:413]: Feedback here would be appreciated @krfricke @xwjiang2010

Comment (xwjiang2010) [reply python/ray/tune/suggest/suggestion.py:421]: I like it!

File: python/ray/tune/suggest/suggestion.py [hunk 2]
```diff
@@ -372,17 +395,41 @@ def __init__(self,
...
+
+    def set_search_properties(self, metric: Optional[str], mode: Optional[str],
+                              config: Dict, **spec) -> bool:
+        self._set_searcher_max_concurrency()
+        return set_search_properties_backwards_compatible(
```
Comment (xwjiang2010) [inline python/ray/tune/suggest/suggestion.py:426]: What is our deprecation cycle? I feel like we can safely remove this already?

File: python/ray/tune/suggest/hebo.py [hunk 3]
```diff
@@ -166,13 +171,19 @@ def __init__(
         self._live_trial_mapping = {}
 
         self._max_concurrent = max_concurrent
+        self._setup_optimizer_again = False
```
Comment (Yard1) [inline python/ray/tune/suggest/hebo.py:174]:
```suggestion
        self._max_concurrent = max_concurrent
```

**Comment by Yard1** [comment]
@xwjiang2010 @krfricke I have updated the PR. I had to break up the `tune_test_restore` into two files, as HEBO was taking a lot longer with concurrency correctly limited. Before, it was using its default mode of 8 concurrent trials despite the test using only one CPU - it didn't impact the test and the side effect of that was that it cached the suggestions, speeding it up a lot. I had fixed that but I left the test broken up as I feel that it makes more sense now, but let me know if you'd rather have it back in one file.

File: python/ray/tune/suggest/ax.py [hunk 4]
```diff
@@ -258,9 +258,9 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
                     metric=self._metric,
                     mode=self._mode))
 
-        if self.max_concurrent:
-            if len(self._live_trial_mapping) >= self.max_concurrent:
-                return None
+        # if self.max_concurrent:
+        #     if len(self._live_trial_mapping) >= self.max_concurrent:
+        #         return None
```
Comment (krfricke) [inline python/ray/tune/suggest/ax.py:263]: Remove?

File: python/ray/tune/suggest/bayesopt.py [hunk 5]
```diff
@@ -245,10 +245,10 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
...
 
         # If we have more active trials than the allowed maximum
-        total_live_trials = len(self._live_trial_mapping)
-        if self.max_concurrent and self.max_concurrent <= total_live_trials:
-            # we stop the suggestion and return None.
-            return None
+        # total_live_trials = len(self._live_trial_mapping)
+        # if self.max_concurrent and self.max_concurrent <= total_live_trials:
+        #     # we stop the suggestion and return None.
+        #     return None
```
Comment (krfricke) [inline python/ray/tune/suggest/bayesopt.py:251]: Remove?

File: python/ray/tune/suggest/hyperopt.py [hunk 6]
```diff
@@ -265,9 +265,9 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
                     metric=self._metric,
                     mode=self._mode))
 
-        if self.max_concurrent:
-            if len(self._live_trial_mapping) >= self.max_concurrent:
-                return None
+        # if self.max_concurrent:
+        #     if len(self._live_trial_mapping) >= self.max_concurrent:
+        #         return None
```
Comment (krfricke) [inline python/ray/tune/suggest/hyperopt.py:270]: Remove?

File: python/ray/tune/suggest/nevergrad.py [hunk 7]
```diff
@@ -240,9 +240,9 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
                     metric=self._metric,
                     mode=self._mode))
 
-        if self.max_concurrent:
-            if len(self._live_trial_mapping) >= self.max_concurrent:
-                return None
+        # if self.max_concurrent:
+        #     if len(self._live_trial_mapping) >= self.max_concurrent:
+        #         return None
```
Comment (krfricke) [inline python/ray/tune/suggest/nevergrad.py:245]: Remove?

File: python/ray/tune/suggest/suggestion.py (hunk 2 above)
Comment (xwjiang2010) [reply python/ray/tune/suggest/suggestion.py:426]: Actually I have just removed it myself :)

**Review by xwjiang2010 — APPROVED** [review]
Thanks Antoni! Yes, let's split up the tests!!

File: python/ray/tune/suggest/hebo.py [hunk 8]
```diff
@@ -219,11 +228,12 @@ def set_search_properties(self, metric: Optional[str], mode: Optional[str],
...
-        space = self.convert_search_space(config)
-        self._space = space
-        if metric:
+        if not self._space:
+            space = self.convert_search_space(config)
+            self._space = space
+        if not self._metric:
             self._metric = metric
-        if mode:
+        if not self._mode:
```
Comment (krfricke) [inline python/ray/tune/suggest/hebo.py:236]: For all other searchers, we have `if mode:` here, in order to always overwrite the mode if passed to `tune.run()`. Should we keep this here, too?

**Review by krfricke — COMMENTED** [review]
Looks good, some question about commented out code paths.

File: python/ray/tune/suggest/hebo.py (hunk 8 above)
Comment (Yard1) [reply python/ray/tune/suggest/hebo.py]: You are right, it should be consistent. Changed

**Pull request merged**